x-pack/lib/filters/geoip/database_manager.rb

```diff
@@ -21,148 +22,202 @@
+  # call expiry action if Logstash use EULA database and fail to touch the endpoint for 30 days in a row
+  def check_age(database_types = DB_TYPES)
+    database_types.map do |database_type|
+      next if !@states[database_type].is_eula || @states[database_type].observable.count_observers == 0
+
+      days_without_update = (::Date.today - ::Time.at(@metadata.check_at(database_type)).to_date).to_i
+
+      case
+      when days_without_update >= 30
+        logger.error("The MaxMind database hasn't been updated from last 30 days. Logstash is unable to get newer version from internet. "\
+          "According to EULA, GeoIP plugin needs to stop using MaxMind database in order to be compliant. "\
+          "Please check the network settings and allow Logstash accesses the internet to download the latest database, "\
+          "or switch to offline mode (:database => PATH_TO_YOUR_DATABASE) to use a self-managed database "\
+          "which you can download from https://dev.maxmind.com/geoip/geoip2/geolite2/ ")
+        @states[database_type].observable.notify_all(:expire)
+      when days_without_update >= 25
+        logger.warn("The MaxMind database hasn't been updated for last #{days_without_update} days. "\
+          "Logstash will fail the GeoIP plugin in #{30 - days_without_update} days. "\
+          "Please check the network settings and allow Logstash accesses the internet to download the latest database ")
+      else
+        logger.trace("The endpoint hasn't updated", :days_without_update => days_without_update)
+      end
+    end
+  end
 
-    begin
-      if execute_download_job
-        @geoip.setup_filter(database_path)
-        clean_up_database
+  # Clean up directories which are not mentioned in metadata and not CC database
+  def clean_up_database
+    protected_dirnames = (@metadata.dirnames + [CC]).uniq
+    existing_dirnames = ::Dir.children(get_data_dir_path)
+                             .select { |f| ::File.directory? ::File.join(get_data_dir_path, f) }
+
+    (existing_dirnames - protected_dirnames).each do |dirname|
+      dir_path = get_dir_path(dirname)
+      FileUtils.rm_r(dir_path)
+      logger.info("#{dir_path} is deleted")
+    end
+  end
+
+  def trigger_download
+    unless @triggered
+      @trigger_lock.synchronize do
+        unless @triggered
+          execute_download_job
+
+          # check database update periodically. trigger `call` method
+          @scheduler = Rufus::Scheduler.new({:max_work_threads => 1})
+          @scheduler.every('24h', self)
+
+          @triggered = true
+        end
       end
-    rescue DatabaseExpiryError => e
-      logger.error(e.message, error_details(e, logger))
-      @geoip.expire_action
     end
   end
 
+  public
+
+  # scheduler callback
+  def call(job, time)
+    logger.debug "scheduler runs database update check"
+    execute_download_job
+  end
+
   def close
     @scheduler.every_jobs.each(&:unschedule) if @scheduler
```

Comment:
Is there a reason not to use [shutdown/stop](https://github.com/jmettraux/rufus-scheduler/blob/master/lib/rufus/scheduler.rb#L509-L534) here instead? 

```suggestion
    @scheduler.shutdown if @scheduler
```

Reply:
I remove this part as no one is calling it